### ISSUE TITLE
B901: Misses `yield` sub-expresisons

### ISSUE THREAD

**Issue by MichaReiser** [astral-sh/ruff#14453]
A very contrived example but `B901` misses yield expressions where they're not the expression of a `ExprStmt`. 

```py
def get_file_paths(file_types: Iterable[str] | None = None) -> Iterable[Path]:
    dir_path = Path(".")
    if file_types is None:
        return dir_path.glob("*")

    for file_type in file_types:
        (yield a for a in dir_path.glob(f"*.{file_type}"))
```

Note: we should make sure that we don't traverse into any lambda expressions...

**Comment by MichaReiser** [comment]
hmm, I only tested with our parser. nested yield expressions are definitely valid :)

**Comment by MichaReiser** [comment]
Okay, found one

```
async def main():
    await (yield x)
```

**Comment by AlexWaygood** [comment]
Here's another function that is a generator function with a `return` statement and is not flagged by B901:

```py
def f():
    x = yield
    print(x)
    return 42
```

This is a generator function that you can `send` values into:

```pycon
>>> def f():
...     x = yield
...     print(x)
...     return 42
...     
>>> y = f()
>>> next(y)
>>> y.send('foo')
foo
Traceback (most recent call last):
  File "<python-input-13>", line 1, in <module>
    y.send('foo')
    ~~~~~~^^^^^^^
StopIteration: 42
```

`Send`ing valued _into_ a generator is, like `return`ing values from a generator, quite an advanced feature. So if I saw a function like this in code review, I would assume the user knew what they were doing. That doesn't feel like a very principled reason not to emit B901 on it, though; maybe we should. Not sure.

**Comment by rabelmervin** [comment]
Hi @MichaReiser @AlexWaygood  can I help you fix this issue ?

**Comment by MichaReiser** [comment]
Sure. You can find the code for this rule in this file

https://github.com/astral-sh/ruff/blob/b7e32b0a18487e8c9813e0d189d7a5d625ff5c3f/crates/ruff_linter/src/rules/flake8_bugbear/rules/return_in_generator.rs

**Comment by rabelmervin** [comment]
Thanks @MichaReiser excited to work on this !

**Comment by MichaReiser** [comment]
We decided to leave this as is for now because there are some valid patterns that would be flagged if we change the rule to detect all yield usages. See the linked PR

**Comment by ntBre** [comment]
I left a longer comment on the PR (https://github.com/astral-sh/ruff/pull/21200#issuecomment-3609003602), but I ended up changing this in #21200 before I found this issue.

I'm happy to revert that if needed, but I'll close this as resolved for now.
